Move constants and required types from 0.1, make `NULL` public

rvolosatovs opened the pull request:
These are (or should be) used by enarx. Haven't checked if everything is actually used.
https://github.com/enarx/enarx is already using 2021
Required for https://github.com/enarx/enarx/issues/1062

Commit:
v2: Mark `NULL` public

This will be used by external call handler implementations

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2: Include constants from 0.1 version of the crate

Removed `MAX_UDP_PACKET_SIZE` constant along the way, since it is not
used by enarx/enarx

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2: Use Rust 2021 edition

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2/item/gdbcall: Reorganize

Define constants/enums after the type and trait definitions for clarity.

Also fixed incorrect ordering of enums - 0x01 comes after 0x00, not 0x02

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2/item/syscall: Add constants and types from 0.1

Removed `virt_start` from `MemInfo`, because the referenced issue is
closed and is fixed by design on the new version.

Signed-off-by: Roman Volosatovs <roman@profian.com>

haraldh commented:
Regarding the removal of `MAX_UDP_PACKET_SIZE`.

Don't we need the wanted size of the block somehow?

```rust
impl Block {
    /// Returns the capacity of `Block.buf`
    pub const fn buf_capacity() -> usize {
        // At least MAX_UDP_PACKET_SIZE rounded up Page::size() alignment
        (MAX_UDP_PACKET_SIZE + size_of::<Message>() + Page::SIZE - 1) / Page::SIZE * Page::SIZE
            - size_of::<Message>()
    }
}
```

sgx `src/backend/sgx/thread.rs`
```
pub struct Thread {
[…]
    block: Block,
[…]
}
```

sev is somehow hardcoded in `internal/shim-sev/layout.ld`
```
_ENARX_SALLYPORT_BLOCK_SIZE = 69632;
```

but later dynamically used with `kvm_builder_map()` in `src/backend/kvm/builder.rs`
using `size_of::<Block>()`

rvolosatovs commented:
I didn't find a usage in https://github.com/enarx/enarx https://github.com/enarx/enarx/search?q=MAX_UDP_PACKET_SIZE, and hence removed it.
The sallyport crate is designed to work with blocks of arbitrary size, and I considered this to be a configuration parameter, value of which is specific to the use case of the consumer of the crate.

If we were to include this constant in the crate, I feel like it would make the most sense to then also provide a `Block` type, which would have a `From` implemented for an arbitrary byte buffer and would validate that the size of that buffer is at least `MAX_UDP_PACKET_SIZE`, although there is nothing within the sallyport crate implementation that actually requires this.

So yeah, I feel like block size is something that should be tightly coupled to the shim implementations, and hence, reside there, but I don't mind including it here either, even if just an exposed constant not actually used by the crate.

Let me know, which approach you think is the most applicable here.

rvolosatovs commented:
How about we provide an API to calculate a "size hint" of the block, given the size of contents you'd want to send?
Then we could also provide the UDP constant and consumers of the crate could do something like `block::size_hint(MAX_UDP_PACKET_SIZE)`, which would return the length of the block required to fit at least one call of biggest size (i.e. a syscall at the moment of writing, which has 7 args) with an allocated `MAX_UDP_PACKET_SIZE` bytes of payload.

Commit:
v2/item: Introduce `Block::size_hint`

This will be used by consumers of the crate to calculate approximate
length of the block to use.
Also move `MAX_UDP_PACKET_SIZE` from 0.1 and add a few tests.
Extracted block-related functionality into a separate submodule.

Signed-off-by: Roman Volosatovs <roman@profian.com>
